spec/minidoc_spec.rb

```diff
@@ -0,0 +1,376 @@
+      expect(user.name).to eq "Noah"
+      expect(user.age).to eq 20
+
+      second_user = SecondUser.find_one(age: 20)
+      expect(user.id).to eq second_user.id
+      second_user.age = 21
+      expect(second_user.age).to eq 21
+      second_user.save
+      expect(second_user.reload.age).to eq 21
+
+      user.reload
+      expect(user.name).to eq "Noah"
+      expect(user.age).to eq 21
+    end
+  end
+
+  describe "#save!" do
+    it "persists the change to the database" do
+      user = User.create!(name: "Bryan")
+      user.name = "Noah"
+      user.save!
+
+      expect(user.reload.name).to eq "Noah"
+    end
+
+    it "does not persist the change to the database when a validation is not satisfied" do
+      expect {
+        ValidationsUser.new.save!
+      }.to raise_error(Minidoc::RecordInvalid)
+    end
+  end
+
+  describe "#valid?" do
+    it "validates" do
+      user = ValidationsUser.new
+      expect(user).to_not be_valid
+
+      user.name = "Noah"
+      expect(user).to be_valid
+    end
+  end
+
+  describe "#persisted?" do
+    let(:user) { User.new }
+
+    it "knows new records are not persisted" do
+      expect(user.persisted?).to eq false
+    end
+
+    it "knows saved records are persisted" do
+      user.save
+      expect(user.persisted?).to eq true
+    end
+
+    it "knows deleted records are not persisted" do
+      user.save!
+      user.destroy
+      expect(user.persisted?).to eq false
+    end
+  end
+
+  describe ".create!" do
+    it "inserts a document into the database" do
+      user = User.create!(name: "Bryan")
+      expect(user.name).to eq "Bryan"
+      expect(user.persisted?).to be true
+      expect(User.count).to eq 1
+    end
+
+    it "raises an error when a validation is not satisfied" do
+      expect { ValidationsUser.create! }.to raise_error(Minidoc::RecordInvalid)
+      expect(User.count).to eq 0
+    end
+  end
+
+  describe ".create" do
+    it "inserts a document into the database" do
+      user = User.create(name: "Bryan")
+      expect(user.name).to eq "Bryan"
+      expect(User.count).to eq 1
+    end
+
+    it "does not insert a document when a validation is not satisfied" do
+      user = ValidationsUser.create
+      expect(user.persisted?).to be false
+      expect(ValidationsUser.count).to eq 0
+    end
+  end
+
+  describe "#destroy, #destroyed?" do
+    it "removes the document from the collection, and the document knows it" do
+      user = User.create!(name: "Bryan")
+      expect(user.destroyed?).to eq false
+
+      user.destroy
+
+      expect(User.count).to eq 0
+      expect(user.destroyed?).to eq true
+    end
+  end
+
+  describe ".delete" do
+    it "removes a document by id" do
+      user = User.create!(name: "Bryan")
+
+      User.delete(user.id)
+
+      expect(User.count).to eq 0
+
+      # FIXME: this is a nice behavior of Mongoid that I'm just realizing Minidoc doesn't have
+      # expect { user.reload }.to raise_error(Minidoc::DocumentNotFoundError)
```

Comment:
Note that this doesn't really belong in this PR but I'm curious if anyone else agrees we should add this behavior to Minidoc. The current behavior is that it fails with an awkward nil error